lgc/util/llpcGfxRegHelper.h

```diff
@@ -0,0 +1,153 @@
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+ /**
+  ***********************************************************************************************************************
+  * @file  llpcGfxRegHelper.h
+  * @brief LLPC header file: contains the definition of LLPC utility class GfxRegHelper.
+  *
+  * @details The class handles symbolic expressions over values read from hardware graphics registers, materialized as
+  *          llvm instructions. The code maintains a map from a register ID to an llvm value with the current symbolic
+  *          expression describing the updated state of the register.
+  *
+  * @sa llpc/docs/DdnGraphicsRegisterHandler.md
+  ***********************************************************************************************************************
+  */
+
+#pragma once
+#include "llpcGfxRegHandler.h"
+#include "llpcTargetInfo.h"
+
+namespace lgc
+{
+
+using namespace llvm;
+
+// =====================================================================================================================
+// The goal of the helper structs is to maintain the newest symbolic state of a register and eventually produce
+// optimized llvm instruction sequences for these states.
+// - pValue : Cache the value which is already being fetched.
+// - isModified : True, if the register is being settled with new value.
+struct BitsState
+{
+    Value* pValue     = nullptr;
+    bool   isModified = false;
+};
+
+// =====================================================================================================================
+// Helper class for handling graphics registers.
+// Note: Don't use GfxRegHelper directly, please implement your own register helper class, such as SqImgSampRegHelper
+class GfxRegHelper : public GfxRegHandlerBase
+{
+protected:
+    GfxRegHelper(Builder* pBuilder, Value* pRegister);
+
+    virtual ~GfxRegHelper() = default;
+
+    // Get register
+    Value* GetRegCommon(uint32_t id);
```

Comment:
This hasn't been resolved. Can you add a quick comment that explains where the ID can come from?
This can be a comment on top of the class instead, since it covers many methods that all take IDs.

Reply:
For all the subclasses, if the only valid IDs come from known enums, maybe make this protected and expose by subclass methods that take enums instead of uint32?

Reply:
> This hasn't been resolved. Can you add a quick comment that explains where the ID can come from?
> This can be a comment on top of the class instead, since it covers many methods that all take IDs.

I've added comment on the top of the class.